From f7db68ec5a3a33afa7a80e09b2394e2d298757ce Mon Sep 17 00:00:00 2001
From: Thao Luong <thao.luong.uw@rvc.renesas.com>
Date: Wed, 13 May 2015 14:37:12 +0700
Subject: [PATCH 2/3] omxaudiodec: workaround for issue: can not decode file
 mp4/aac or file m4a

omxaacdec can decode file.aac (adts format) but it can not decode
file.mp4/m4a (raw format)
Reason: in gstreamer1.2.0 audiodec has handle for codec data got from
cap however Renesas audio MC (aacdec) does not support handling codec
data in raw format.
So when gstreamer sends codec data to MC, it will notify error
Workaround: Add hacks (not-handle-codec-data) to skip geting codec data
from cap and send it to MC when using Renesas MC.

refs #69205
Signed-off-by: Thao Luong <thao.luong.uw@rvc.renesas.com>
---
 config/rcar/gstomx.conf |  2 +-
 omx/gstomx.c            |  3 +++
 omx/gstomx.h            |  7 +++++++
 omx/gstomxaudiodec.c    | 26 ++++++++++++++++----------
 4 files changed, 27 insertions(+), 11 deletions(-)

diff --git a/config/rcar/gstomx.conf b/config/rcar/gstomx.conf
index 4e9aef9..fd4b514 100644
--- a/config/rcar/gstomx.conf
+++ b/config/rcar/gstomx.conf
@@ -5,4 +5,4 @@ component-name=OMX.RENESAS.AUDIO.DECODER.AAC
 rank=256
 in-port-index=0
 out-port-index=1
-hacks=no-disable-outport
+hacks=no-disable-outport;not-handle-codec-data
diff --git a/omx/gstomx.c b/omx/gstomx.c
index 9a77e5b..faa4c4a 100644
--- a/omx/gstomx.c
+++ b/omx/gstomx.c
@@ -3,6 +3,7 @@
  *   Author: Sebastian Dröge <sebastian.droege@collabora.co.uk>, Collabora Ltd.
  * Copyright (C) 2013, Collabora Ltd.
  *   Author: Sebastian Dröge <sebastian.droege@collabora.co.uk>
+ * Copyright (C) 2015, Renesas Electronics Corporation
  *
  * This library is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -2459,6 +2460,8 @@ gst_omx_parse_hacks (gchar ** hacks)
       hacks_flags |= GST_OMX_HACK_NO_COMPONENT_ROLE;
     else if (g_str_equal (*hacks, "no-disable-outport"))
       hacks_flags |= GST_OMX_HACK_NO_DISABLE_OUTPORT;
+    else if (g_str_equal (*hacks, "not-handle-codec-data"))
+      hacks_flags |= GST_OMX_HACK_NOT_HANDLE_CODEC_DATA;
     else
       GST_WARNING ("Unknown hack: %s", *hacks);
     hacks++;
diff --git a/omx/gstomx.h b/omx/gstomx.h
index 9ad8e2f..90789ae 100644
--- a/omx/gstomx.h
+++ b/omx/gstomx.h
@@ -3,6 +3,7 @@
  *   Author: Sebastian Dröge <sebastian.droege@collabora.co.uk>, Collabora Ltd.
  * Copyright (C) 2013, Collabora Ltd.
  *   Author: Sebastian Dröge <sebastian.droege@collabora.co.uk>
+ * Copyright (C) 2015, Renesas Electronics Corporation
  *
  * This library is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -116,6 +117,12 @@ G_BEGIN_DECLS
  */
 #define GST_OMX_HACK_NO_DISABLE_OUTPORT                               G_GUINT64_CONSTANT (0x0000000000000100)
 
+/* Renesas audio MC (aacdec) does not support handling codec data with 
+ * current input formats.
+ * So this hack will skip collecting and handling codec data.
+ */
+#define GST_OMX_HACK_NOT_HANDLE_CODEC_DATA                            G_GUINT64_CONSTANT (0x0000000000000200)
+
 typedef struct _GstOMXCore GstOMXCore;
 typedef struct _GstOMXPort GstOMXPort;
 typedef enum _GstOMXPortDirection GstOMXPortDirection;
diff --git a/omx/gstomxaudiodec.c b/omx/gstomxaudiodec.c
index f894086..e35af30 100644
--- a/omx/gstomxaudiodec.c
+++ b/omx/gstomxaudiodec.c
@@ -4,6 +4,8 @@
  * Copyright (C) 2013, Collabora Ltd.
  *   Author: Sebastian Dröge <sebastian.droege@collabora.co.uk>
  * Copyright (C) 2014, Sebastian Dröge <sebastian@centricular.com>
+ * Copyright (C) 2015, Renesas Electronics Corporation
+ * 
  *
  * This library is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -831,16 +833,20 @@ gst_omx_audio_dec_set_format (GstAudioDecoder * decoder, GstCaps * caps)
           NULL) != OMX_ErrorNone)
     return FALSE;
 
-  /* Get codec data from caps */
-  gst_buffer_replace (&self->codec_data, NULL);
-  s = gst_caps_get_structure (caps, 0);
-  codec_data = gst_structure_get_value (s, "codec_data");
-  if (codec_data) {
-    /* Vorbis and some other codecs have multiple buffers in
-     * the stream-header field */
-    self->codec_data = gst_value_get_buffer (codec_data);
-    if (self->codec_data)
-      gst_buffer_ref (self->codec_data);
+  /* Do not get codec data from cap if use "not-handle-codec-data" hack */
+  if (! (klass->cdata.hacks & GST_OMX_HACK_NOT_HANDLE_CODEC_DATA))
+  {
+    /* Get codec data from caps */
+    gst_buffer_replace (&self->codec_data, NULL);
+    s = gst_caps_get_structure (caps, 0);
+    codec_data = gst_structure_get_value (s, "codec_data");
+    if (codec_data) {
+      /* Vorbis and some other codecs have multiple buffers in
+       * the stream-header field */
+      self->codec_data = gst_value_get_buffer (codec_data);
+      if (self->codec_data)
+        gst_buffer_ref (self->codec_data);
+    }
   }
 
   GST_DEBUG_OBJECT (self, "Enabling component");
-- 
2.4.0

